File: scala/src/main/scala/openchat/application/handlers/LoginHandler.scala
package openchat.application.handlers

import openchat.application.Handler
import openchat.application.repositories.UserRepository

import scala.concurrent.{ExecutionContext, Future}

final case class LoginUser(username: String, password: String)
final case class UserLoggedIn(id: String, username: String, about: String)

final class LoginHandler(userRepository: UserRepository)(implicit ec: ExecutionContext)
    extends Handler[LoginUser, Option[UserLoggedIn]] {
  def handle(loginUser: LoginUser): Future[Option[UserLoggedIn]] = {
    userRepository.getByUsername(loginUser.username).map {
      case None       => None
      case Some(user) =>
        if (user.login(loginUser.password))
          Some(UserLoggedIn(id = user.userId, username = user.username, about = user.about))
        else None
    }
  }
}

File: typescript/src/application/handlers/LoginHandler.ts
import type { Handler } from '../Handler'
import type { UserRepository } from '../repositories/UserRepository'
import type { UserId } from '../../domain/aggregates/User'

export class LoginHandler implements Handler<LoginUser, UserLoggedIn> {
  constructor(private userRepository: UserRepository) {}

  handle = async (loginUser: LoginUser): Promise<UserLoggedIn> => {
    const user = await this.userRepository.getByUsername(loginUser.username)
    if (user === null) return null

    return user.login(loginUser.password)
      ? {
          id: user.userId,
          username: user.username,
          about: user.about,
        }
      : null
  }
}

export type LoginUser = {
  username: string
  password: string
}

export type UserLoggedIn = {
  id: UserId
  username: string
  about: string
} | null
